text/0114-resolved-ts-for-large-transactions.md

```diff
@@ -0,0 +1,205 @@
+
+In current TiKV(v8.3), large transactions can block resolve-ts from advancing, because it is calculated as `min(pd-tso, min(lock.ts))`, which is actually a more stringent constraint than its original definition. A lock from a pipelined txn can live several hours. This will make services dependent on resolved-ts unavailable.
+
+## Goals
+
+- Current phase objectives:
+  - Prevent pipelined transactions from impeding resolved-ts advancement.
+
+- Long-term goal:
+  - Minimize the overhead of resolved-ts maintenance.
+  - Maximize resolved-ts freshness.
+  - Achieve uninterrupted resolved-ts progression, addressing all potential blocking factors beyond long transactions and their associated locks. (Further details to be discussed in the final section of this proposal.)
+
+## Design
+
+Key Concept: Utilizing `lock.min_commit_ts` instead of `lock.start_ts` for resolved-ts calculation.
+
+Rationale:
+1. Resolved-ts definition is independent of LOCK CF.
+2. Current use of `lock.start_ts` is based on the invariant: `lock.start_ts` < `lock.commit_ts`.
+3. A valid resolved-ts doesn't necessitate the absence of locks with smaller start_ts, provided their future commit_ts are guaranteed to be larger.
+
+Advantages of `lock.min_commit_ts`:
+1. Satisfies a similar invariant: `lock.min_commit_ts` <= `lock.commit_ts`
+2. Unlike the static `lock.start_ts`, `lock.min_commit_ts` can be dynamically increased.
+
+### Maintanence of resolved-ts
+
+Key objective: Maximize all TiKV nodes' awareness of large pipelined transactions during their lifetime, i.e. from their first writes to all locks being committed. These info are necessary:
+
+1. `start_ts`
+2. A fresh enough `min_commit_ts`
+3. Status
+
+#### Coordinator
+
+For a pipelined transaction, its TTL manager is responsible for fetching a latest TSO as a candidate of min_commit_ts and update both the committer's inner state and PK in TiKV. After that, it broadcasts the `start_ts` and the new `min_commit_ts` to all TiKV stores. The PK update can be piggybacked on the `heartbeat` request.
+
+Optionally, to avoid too many RPC overhead, the broadcast messages from different transactions can be batched.
+
+Atomic variables or locks may be needed for synchronization between the TTL manager and the committer.
+
+#### Scaling out TiKVs
+
+Challenge:
+During cluster expansion, when a new TiKV instance is integrated mid-transaction, the TTL manager must promptly incorporate it into its broadcast list. The TTL manager relies on the region cache for store information. However, if the region cache lacks awareness of a newly added TiKV, the TTL manager may inadvertently omit it from broadcasts.
+
+One solution is to have an background goroutine in the region cache to periodically refresh the complete store list.
+
+#### TiKV scheduler - heartbeat
+
+Besides updating TTL, it also supports update min_commit_ts of the PK.
+
+*TBD: should it also update max_ts?*
+
+#### TiKV txn_status_cache
```

Comment:
Does it mean that it requires all information about these large txns must exist in the cache? If so, I'm afraid it's no longer a *cache* as literal.

Reply:
The cache operates on a "best effort" basis rather than being mandatory. In the case of cache miss, the system falls back to use start_ts.